README.md

```diff
@@ -64,11 +64,86 @@ Neon's components need to react to various events:
+A minimal integration can be achieved by configuring a `Highlighter` to interface with an `NSTextView`'s text container:
+
+```swift
+func applicationDidFinishLaunching(_ aNotification: Notification) {
+	guard let textContainer = textView.textContainer, let textStorage = textView.textStorage else {
+		preconditionFailure()
+	}
+	let textInterface = TextContainerSystemInterface(textContainer: textContainer, attributeProvider: self.attributeProvider)
+	self.highlighter = Highlighter(textInterface: textInterface, tokenProvider: self.tokenProvider)
+	textStorage.delegate = self
+	self.highlighter.invalidate()
+}
+```
+
+Attaching the highlighter to a text view interface tells it _what_ to update, but not _when_. You have to notify the highlighter whenever the text view's content changes, and invalidate existing highlighting as needed. Such notifications can be conveyed by making yourself the delegate of your text view's `textStorage`, and implementing this delegate method:
+
+```swift
+func textStorage(_ textStorage: NSTextStorage, didProcessEditing editedMask: NSTextStorageEditActions, range editedRange: NSRange, changeInLength delta: Int) {
+
+	// Map NSTextStorageDelegate editedRange to Neon's style of editedRange
+	let adjustedRange = NSRange(location: editedRange.location, length: editedRange.length - delta)
+	self.highlighter.didChangeContent(in: adjustedRange, delta: delta)
+
+	DispatchQueue.main.async {
+		self.highlighter.invalidate()
+	}
+}
+```
+
+Notice that the `invalidate` method is dispatched asynchronously to ensure the styles are not updated until after the underlying text storage is done being edited.
+
+The initial configuration of `highlighter` included references to `self.tokenProvider` and `self.attributeProvider`, which are responsibility for providing the logic behind _what_ gets highlighted, and _how_ it should be done. At a minimum, the `TokenProvider` generates and supplies named tokens that correspond to ranges of text:
+
+```swift
+let paintItBlackTokenName = "paintItBlack"
+
+func tokenProvider(_ range: NSRange, completionHandler: @escaping (Result<TokenApplication, Error>) -> Void) {
+	var tokens: [Token] = []
+
+	if let searchString = self.textView.textStorage?.string {
+		if let nonWhitespaceRegex = try? NSRegularExpression(pattern: "[^\\s]+\\s{0,1}") {
+			nonWhitespaceRegex.enumerateMatches(in: searchString, range: range) { regexResult, _, _ in
+				guard let result = regexResult else { return }
+				for rangeIndex in 0..<result.numberOfRanges {
+					let tokenRange = result.range(at: rangeIndex)
+					tokens.append(Token(name: paintItBlackTokenName, range: tokenRange))
+				}
+			}
+		}
+	}
```

Comment:
This is totally up to you. But, you could employ `guard`, just like you do within the `enumerateMatches` to reduce the indentation here quite a bit. Could be nice for smaller screens!

Reply:
The reason I didn't here was to avoid a redundant completionHandler call, but maybe that would be preferable to tame the indenting.

Reply:
I ended up taking this advice and the current branch has a guard statement.